### PR TITLE
WIP Add functional front-end testing

### PR DESCRIPTION
**Pull request by IshaVDave**
Purpose:
Testing various features of the web app layout-web-tool using Selenium Webdriver and Flask-testing.

### CODE REVIEW

**Comment by IshaVDave** [comment]
A couple things:

1. There are some .vs files that got included into the PR and I'm not sure if you need them (they are meant only for some Visual Studio functionality right?). Once that is confirmed, I will remove them.
2. The code still doesn't work, and I'm in the process of getting it to.

File: .vs/ProjectSettings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "CurrentProjectSetting": null
+}
```
Comment (SantiagoTorres) [inline .vs/ProjectSettings.json]: I think all of the .vs files shouldn't be checked in, as they are configuration settings from your editor. You probably want to add
```
.vs/*
```
to our gitignore so we don't add it in further commits. For now you want to remove these files from the commit and update the pr...

File: tests/selenium_testing.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    driver = self.driver
+    driver.get('https://in-toto.engineering.nyu.edu/')
+
+    start = driver.find_element_by_xpath('/html/body/div[3]/div/div/a')
+    start.click()
+
+    self.driver.quit()
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (SantiagoTorres) [inline tests/selenium_testing.py]: In order for this to work, you need to add these requirements to `requirements.txt`

```
diff --git a/requirements.txt b/requirements.txt
index f23a780..26d84e5 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -2,3 +2,5 @@ Flask==1.1.2
 Flask-PyMongo==2.3.0
 Flask-WTF==0.14.3
 in-toto==0.5.0
+Flask-Testing
+selenium
```

**Review by SantiagoTorres — CHANGES_REQUESTED** [review]
Pretty cool! 

selenium
I think we need a couple of minor tweaks before merging...

**Comment by IshaVDave** [comment]
Thanks for the review! I have made the changes and everything should work now. Two more things:
1. Should I be rebasing the *selenium-testing file* and *updating gitignore and requirements.txt* into two commits for better history?
2. Should I keep adding tests or should it be a separate PR?

**Comment by lukpueh** [comment]
> Thanks for the review! I have made the changes and everything should work now. Two more things:
> 
> 1. Should I be rebasing the _selenium-testing file_ and _updating gitignore and requirements.txt_ into two commits for better history?

IMO you can squash all 3 commits into one and add a nice description of your addition.

> 2. Should I keep adding tests or should it be a separate PR?

I'm fine either way. IIUC the PR doesn't test anything yet, right? Maybe you can add one initial basic test case to demonstrate how selenium works.

File: tests/selenium_testing.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
+"""
+  Setting up Functional testing for layout web tool.
```
Comment (lukpueh) [inline tests/selenium_testing.py]:
```suggestion
  Setting up functional testing for layout web tool.
```

File: tests/selenium_testing.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      # Default port is 5000
+      app.config['LIVESERVER_PORT'] = 8943
+      # Default timeout is 5 seconds
+      app.config['LIVESERVER_TIMEOUT'] = 10
+      return app
+
+  def setUp(self):
+    self.driver = webdriver.Chrome()
+
+  def test_Start(self):
```
Comment (lukpueh) [inline tests/selenium_testing.py]:
```suggestion
  def test_start(self):
```

**Review by lukpueh — COMMENTED** [review]
Very cool stuff, @IshaDave! Apart from two minor nits inline this looks good to go. 

It would also be nice to have a bit more code documentation, or an exemplary test case to make it easier for someone who wants to add tests. What do you think?

**Comment by IshaVDave** [comment]
> IMO you can squash all 3 commits into one and add a nice description of your addition.

Will do!
> IIUC the PR doesn't test anything yet, right?

So a primary test that @SantiagoTorres and I had discussed was to see if the website opened and return a screenshot of the page, then check if you can proceed to the next and come back to this page, then return another screenshot. If the two screenshots are the same, the test runs successfully. I might have misunderstood that part of the instruction, but I am not able to find a way out to compare those two screenshots.

**Comment by adityasaky** [comment]
@SantiagoTorres indicated this PR is ready for to be merged, @IshaDave. Could I ask you to squash some commits before we proceed? :smile:

**Comment by IshaVDave** [comment]
> @SantiagoTorres indicated this PR is ready for to be merged, @IshaDave. Could I ask you to squash some commits before we proceed? 😄

Sounds great and will do!

**Comment by adityasaky** [comment]
Thanks for the squashes, @IshaDave!

**Pull request merged**